connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectWorkerIntegrationTest.java

```diff
@@ -1123,6 +1134,137 @@ public void testTasksMaxEnforcement() throws Exception {
         );
     }
 
+    @Test
+    public void testKafka16838() throws Exception {
```

Comment:
This test is a bit ugly and may be overkill for these small bugs. I've left it in in order to demonstrate that both of the bugs addressed in this PR can be reproduced end-to-end.

If this test seems worth including, I can try to clean it up (add a more descriptive name, possibly break out into two separate tests, and possibly find a more elegant way to force topic compaction).

And of course, if this test doesn't seem worth adding to our integration test suite, I'm happy to remove it.

Reply:
Yep, done. I've split it out into two cases as well.